### ISSUE TITLE
Thoughts on TimeTest

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#2237]
Defects & complexities in TimeTest have been causing discussion, both as Issues and PRs recently and
for years.  Most recently PR #2235 suggests a fix to `localtime` portability issues.

The issue that PR addresses is that the two `localtime` tests fail when run outside of PR and
on some systems in the wild, particularly when Daylight Saving Time is active. Since DST
changes in the background, the defect becomes manifest or not as the year progresses.

That PR identified a key section of the Posix 2018 specification, that the variable timezone is
relative to the 'standard' timezone,  From some private design studies, I think that part of the
specification can be used to make the localtime tests useful in most, but not all, locations.

Because of the underlying complexities of time handling, I think that we should aim for 
correct and workperson-like quality, but not try to cover all cases. That is, apply the Pareto
(70/30, 80/20, etc. rule).   No matter what we do, we are going to be back here again.

To work:

   The `tm_isdst` field is part of the Posix specification of `struct tm`. `localtime` and `localtime_r`
   are supposed to fill that field with one of three values: -1 Unknown, 0 Standard, 1 Daylight.

   The second key insight is that the Unix Epoch was Standard time in most of the Earth (Duh!)
   and should be reported as such.  (See next comment in this Issue for details & discussion.)
  
   `localtime()` and `localtime_r()` should have all the information they need from the environment
   in order to set `tm_isdst` because the TimeTest class constructor explicitly calls `tzset()`. 
   
  These two points taken together with the `timezone` requirement identified by PR #2235
    `timezone` can be presumed valid in a known set of circumstances.  

    1)  is_dst == -1   fail() test 
         This will at least identify the local conditions causing problems.

    2) is_dst  == 1  use assumeFalse() to give message and ignore
        rest of the test.   

        This means that running the test in large parts of Earth's southern
        hemisphere will log a scary message and yield no value.   
  
        While I strive for end-to-end tests that work everywhere, I think
        that focusing on the purpose of the TimeTest localtime* tests
        is wise. That purpose is to test if the "glue layer" entry point and structure
        handling is correct, not if the underlying libc daylight saving time
        handling is correct.  The latter would be a comforting bonus, but
        the added complexity is more likely to add bugs than increased
        confidence.
 
     3)  is_dst == 0.  Use addition of timezone.

**Comment by LeeTibbert** [comment]
Re: 
```The second key insight is that the Unix Epoch was Standard time in most of the Earth (Duh!)
and should be reported as such.```

To dive into complexity, where at least one of the bugs we seek may be hidden

1) The `localtime` tests have been running, apparently successfully, in a number of environments
    outside of CI. Those environments have been both Standard and Daylight Saving time. 
    I think that fact is telling us something. Probably that there is a ~~weird~~ unexpected/un-handled
    environment in play where the bug appears.

2) I did a short design study to test my expectation that Unix Epoch gets reported as "tm_isDst == 0",
    at least on my system which is currently operating with Daylight Saving Time. Yes, it get reported
    as the expected zero.  Crude code:

````
  @Test def localtimeShouldTransformTheEpochToLocaltime(): Unit = {
    val time_ptr = stackalloc[time_t]
    !time_ptr = epoch + timezone()
    val time: Ptr[tm] = localtime(time_ptr)
    val cstr: CString = asctime(time)
    val str: String   = fromCString(cstr)

printf(s"\n>>> LeetDebug: time.tm_isdst: |${time.tm_isdst}|\n")
printf(s">>> LeetDebug: timezone: |${timezone}|\n\n ")

    assertTrue("Thu Jan  1 00:00:00 1970\n".equals(str))
  }

```

@ncordon Could you run that code, or something similar to it on your system.
Perhaps there is some OS difference (egads!  What is the French expression "quelle suprise!".)

My Ubuntu 20.10 glibc system may be doing the "expected" "is_dst at the time of the supplied time_t"
and yours may be doing "is_dst at current time".

**Comment by LeeTibbert** [comment]
I have also noticed that all of the tests in TimeTest are cunningly set up to use times
which are highly unlikely to be "is_dst == 1" in Northern Hemisphere.  I guess that
case gets exercised, so far without problems, in the Southern Hemisphere.

**Comment by LeeTibbert** [comment]
I also understand if someone appeals to the Pareto principle and the huge backlog of
other issues and suggests that the two problematic `localtime` tests become
"Continuous Integration" only, so that they run in a controlled, contained, known
environment.

That would require the unit-tests ContinuousIntegrationEnviornment PR I am 
creating to allow other work to advance.

It all depends upon how much patience people have.

**Comment by ncordon** [comment]
> @ncordon If you have time & patience, could you run that code, or something
similar to it on your system.

I've run this directly in C:
```c
#include <time.h>
#include <stdio.h>

int main() {
    tzset();
    int epoch = 0;
    time_t rawtime = epoch;
    struct tm *timeinfo;

    timeinfo = localtime(&rawtime);
    printf("is_dst has value %i \n", timeinfo->tm_isdst);
    printf("%s \n", asctime(timeinfo));
}
```

which outputs (in two different OSs):

```
is_dst has value 0
Thu Jan  1 01:00:00 1970
```

so it's not a bug in the Scala native integration and I've confirmed `is_dst = 0`. And you see what `asctime` returns me has the UTC+1 format.

**Comment by LeeTibbert** [comment]
@ncordon Thank you for the C output, it helps.  The `is_dst ==0`,
is what I was expecting/hoping_for.   

 Could you print out the value of the  C `timezone` variable?   I am expecting
it to be -3660 but it is probably 0.

Do you mind telling, here or in a private message, which two operating systems?

Re: Proposal.   Let me let it sit on my back brain.  

      1) I would have to check the  Olson files, but I am pretty sure that there are a couple of places
          (Arizona?) where there is a half-hour change.  That is, that the "1-hour" assumption
          is tempting, but fails often enough that you/we will be fielding bug reports
          for the rest of our professional lives.  That would actually be goodness,
          because it would mean that Scala Native has seen wide adoption in the wild.

     2)  I think that `localtime_r`` does not break circularity. In general, I like the `_r` 
          functions. IIRC localtime_r has its own set of requirements (i.e. not setting
         tzname() and needing to check if a name string is set before the
        timezone variable can be used.)  I think they are covered in TimeTest.scala
        by the explicit `tzset()` call, but they are a field filled with pitfalls.

I'd like to find out why what we are seeing on two different systems is so different
but I am well into thinking  that the cost effective solution is to make the two
tests CI only. 

I still can not let go yet of the forlorn hope of finding the difference.  Monkey 
in the coconut trap.  Can't let go of a fistful of desirable coconut in order
to gain its freedom.

**Comment by LeeTibbert** [comment]
Not a definitive source, but perhaps useful:

[timezone deltas](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)

From inspecting the differences between standard and DST until my eyes glazed over
all the deltas appear to be either 0 or +/- 1 hour.  

A machine scan of the tzdata files might confirm but would be costly.

[historical odd changes] (https://en.wikipedia.org/wiki/Daylight_saving_time)
A one-hour change is customary, but twenty-minute and two-hour changes have been used in the past.

Interrupted by Real Life, gotta go

**Comment by ncordon** [comment]
I scrapped the place today (gist to the Ammonite script [here](https://gist.github.com/ncordon/3311003e49708ac4485a76bcbeedeab8)).

Unfortunately, I'm seeing a few places that do not operate on normal assumptions (offset of +60 minutes in dst):

```
Africa/Casablanca has an offset of -60
Africa/El_Aaiun has an offset of -60
Antarctica/Troll has an offset of 120
Australia/Lord_Howe has an offset of 30
Europe/Dublin has an offset of -60
```

So I think it's best to execute those only on a CI controlled environment. Also I don't think `is_dst` can be entirely trusted on OSX. It's returning 1 (the correct one) or 0 for me depending on the execution of the C program:

```c
#include <time.h>
#include <stdio.h>

int main() {
    tzset();
    time_t rawtime;
    struct tm timeinfo;

    localtime_r(&rawtime, &timeinfo);
    printf("is_dst has value %i \n", timeinfo.tm_isdst);
    printf("timezone has value %li \n", timezone);
    printf("%s \n", asctime(&timeinfo));
}
```

```
is_dst has value 1
timezone has value 0
Mon Aug  5 04:57:25 2120

is_dst has value 1
timezone has value 0
Mon Sep  7 21:52:53 2116

is_dst has value 1
timezone has value 0
Mon Apr  7 18:17:25 2110

is_dst has value 0
timezone has value 0
Sat Feb 27 12:18:45 2106

is_dst has value 1
timezone has value 0
Sun Aug 22 19:55:33 2117

is_dst has value 0
timezone has value 0
Tue Nov  2 06:41:41 2117
```

I have an Arch Linux and a OSX

**Comment by LeeTibbert** [comment]
@ncordon  Sorry to be terse. I hope to have more time after the long holiday weekend.

1) I concur that the `localtime*()` tests should be CI-only, if anybody is ever going
     to get some peace. 

     PRs which contain code which is never used "dead-code" are frowned upon.
     My `java.nio.file.attribute` PR which also needs CI-only is not going to be
      ready in the next few days.  I created  Work In Progress PR #2241 
      to show what I have been working on to the community.

      The `ContinuousIntegration.Scala` file (or an improvement of it) 
      would need to be bundled with either a TimeTest PR or my
      `file.attributes` PR.  I suspect that you are likely to get over
      the goal line first, so feel free to use the code of the WIP 
      PR or improvements.  I am pretty sure the WIP PR itself
      will never be promoted (because it has unity of concept
      but is "dead code" now.)

**Comment by LeeTibbert** [comment]
@ncordon To test out my understanding of the conditions you are using 
and for people reading this topic in the future (may their pain subside over the eons).


Your  "standard", in the non-Irish sense, time is UTC.  You "daylight" time is UTC+1?

**Comment by LeeTibbert** [comment]
I think at one point in this Issue thread there was some C output posted
about macOS and flaky/intermittent results.  Sorry, I can not find it now,
consequences of Evil Living (or Real Life)...

I believe the output was from the next block or similar
```
 timeinfo = localtime(&rawtime);
    printf("is_dst has value %i \n", timeinfo->tm_isdst);
    printf("%s \n", asctime(timeinfo));
}
``` 

I am aware that this is hasty debug code (and may have written it
myslef).
The varying results on macOS may have come from localtime()
failing (returning NULL) and that return code not being checked.
Stack garbage would, I believe, get printed out.

What I saw in the log sure looked like j-random stack garbage.

Now, that leaves open the question as to why localtime() would
be intermittently failing on macOS.  Perhaps code where
the return value is checked would tell us more.

Heisenbugs are always so much fun.

**Comment by ncordon** [comment]
> Your "standard", in the non-Irish sense, time is UTC. You "daylight" time is UTC+1?
Yes, that's correct.

> I think at one point in this Issue thread there was some C output posted
about macOS and flaky/intermittent results

Ah yeah, ignore that, I was doing something wrong. I was not initializing the time with now (so of course results were random 😂). I edited the message 20 minutes after posting it to avoid confusion.

**Comment by LeeTibbert** [comment]
Resolved by the merge of PR #2235.  Thank you @ncordon.
